Replacing modules

LnL7 opened the pull request:
###### Motivation for this change

Enables replacing existing modules, making it easier to use an updated module while still using a stable release for the rest of the system. Example: https://gist.github.com/LnL7/bd05394218033cff17ec20eb93a44a4a

/cc @nbp

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

basvandijk commented:
Very good to see this! I wished this existed several times in the past.

 Since this mechanism is more general than replacing a module (it really just excludes a module), wouldn't it be better to substitute the name `replaces` with `excludes`? Whether you use it for replacing a module is up to the user.

You can then optionally build a very light abstraction on top of this where the replacement is more obvious:

```nix
{
  replaces = [
    { old = <nixpkgs/nixos/modules/services/databases/postgresql.nix>;
      new = ./postgresql.nix;
    }
  ];
}
```

LnL7 commented:
I didn't really give a good example, but the idea is that a custom module can declare what modules it replaces.

basvandijk commented:
Ok you mean placing the `replaces = [ ... ];` in `./postgresql.nix` itself. Then the name makes more sense indeed. I guess I would still prefer a name which describes what operational effect it has instead of the intended application. However, I don't care that much about it. 

Regardless of name, it would be cool to have this feature!

edolstra commented:
Using a fully qualified path to denote modules is unsound because there is no guarantee that '<nixpkgs>' points at the Nixpkgs tree used in the current evaluation. For example, if I do
```
nix-build /my/nixpkgs/nixos/tests/foo.nix -A test
```
then `<nixpkgs>` does not point at `/my/nixpkgs`. Similarly, in a multi-machine NixOps network spec, each machine could use a different Nixpkgs tree. (I don't think NixOps actually already supports that, but you get the idea.)

It might be better to have something that uses module filenames relative to the current Nixpkgs tree, e.g.
```
disableModules = [ "nixos/modules/services/databases/postgresql.nix" ];
```
(I wouldn't call it `replaces` because it doesn't actually replace a module.) The path of the current machine's Nixpkgs tree is available via `config.nixpkgs.path` IIRC.

LnL7 commented:
I agree that it's not ideal to use the full path, but there doesn't seem to be reliable way get the path of `pkgs`, `config.nix.nixPath` doesn't update as far as I can tell.

I also noticed this behaviour, is that expected?
```
$ nix-instantiate --eval --strict -I nixpkgs=/tmp/foo -E '<nixpkgs>'  # /tmp/foo
$ nix-instantiate --eval --strict -I nixpkgs=/tmp/foo -E '<nixpkgs/nixos>'  # /nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs/nixos
```

LnL7 committed:
modules: add support for module replacement with disabledModules

This is based on a prototype Nicolas B. Pierron worked on during a
discussion we had at FOSDEM.

LnL7 commented:
I think this will disable the correct modules when `<nixpkgs>` is not the active source of the nixos modules.

basvandijk commented on lib/modules.nix:
```diff
@@ -87,6 +84,15 @@ rec {
     in result;
 
+
+ /* Remove modules if there is another module that 'disables' it it. */
```
Maybe use something like this: "Return the non-disabled modules. Disabling modules allows replacing their implementations."

basvandijk commented on lib/modules.nix:
```diff
@@ -87,6 +84,15 @@ rec {
+ filterModules = modules:
+   let
+     moduleKey = m: if isString m then toString ../nixos/modules + "/" + m else toString m;
+     replacedKeys = map moduleKey (concatMap (m: m.disabledModules) modules);
```
I would call `replacedKeys` `disabledKeys`.

LnL7 replied:
yeah, that makes more sense.

basvandijk reviewed (changes requested):
Smart trick to just use: `toString ../nixos/modules + "/" + m`.

nbp commented on lib/modules.nix:
```diff
@@ -87,6 +84,16 @@ rec {
+    their implementation to be replaced. */
+ filterModules = modules:
+   let
+     moduleKey = m: if isString m then toString ../nixos/modules + "/" + m else toString m;
```
`../nixos/modules` suppose that all modules that would be replaced would be in NixOS repository.
I honestly does not like that because this gives option to NixOS which are not available to others.

This does not support the idea that someone else might want to use the module system for something else, such as NixUP, or extend NixOS with an additional set of default modules.

LnL7 replied:
That's why I left in the option to use full paths, or are  you referring to the `nixos/modules` prefix?

nbp replied:
My point is that I do not understand why NixOS should be any different than any other consumers of the module system.

nbp replied:
If you can find a way such that one can move this default path as an argument of the module system instead of as a static string in it, this might be preferable.

An example, even without leaving NixOS, is for sub-modules which are defined in different paths, should we name the complete NixOS module path within NixOS, or just name it as a relative file, as this trick is doing for NixOS modules.

LnL7 replied:
I just found [`specialArgs.modulesPath`](https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/eval-config.nix#L52). If we used that other module sets would get the same behaviour if they define define it there.

nbp commented on lib/modules.nix:
```diff
@@ -87,6 +84,16 @@ rec {
 
+
+ /* Filter disabled modules. Modules can be disabled allowing
+    their implementation to be replaced. */
```
nit: use `#` comments, as done above.

nbp commented on lib/modules.nix:
```diff
@@ -87,6 +84,16 @@ rec {
+ filterModules = modules:
+   let
+     moduleKey = m: if isString m then toString ../nixos/modules + "/" + m else toString m;
+     disabledKeys = map moduleKey (concatMap (m: m.disabledModules) modules);
```
Design question, before accepting this change.  Do we want to exclude a single module, or do we want to exclude a single module and all the modules that are inherited by it?

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,36 @@
+
+  options = {
+    services.postgresql.enable = mkOption {
+      description = "Replaced option.";
```
This example is not working as an option should requires more information to be listed.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,36 @@
+<title>Replace Modules</title>
+
+<para>It's also possible to disable modules allowing the
+  implementation to be replaced.  This can be used to import a set of
```
Can you clarify what does it mean. Is that supposed to replace only all option definitions?

Note that `implementation` is still used in many modules as a comment above the `config` attribute, as the `interface` is used above the `options` attributes.  The proper names are `options definitions` and `option declarations`, respectively for `config` and `options` attribute sets.

nbp reviewed (changes requested):
I wish you would have pinged me to push my work on github or at least give some credit back, referring to our discussion at the FOSDEM.

Apart from that, there is at least a design question which needs to be answer, which is: should we consider the closure of the excluded modules or only a single module at a time?

nbp commented:
@edolstra 
> The path of the current machine's Nixpkgs tree is available via `config.nixpkgs.path` IIRC.

No, this is not an option, because this code solve the closure of the modules considered for the module system.  So `config.[...]` is doomed to do an infinite recursion.  This can only leave as a special attribute of the module system, and cannot be influenced by any mean by the evaluation of the configuration, like the imports.

LnL7 commented:
@nbp Sorry, I mainly wanted to get some discussion started around this because I think this would be very nice to have in 17.03. Feel free to close this and push your changes if you want.

nbp commented:
@LnL7 , also can you add tests cases for checking that `disabledModules` allow some to declare an option which would otherwise conflict, and to check that we can remove a definition as well.  There is a test directory under `lib` which already contains module test that you can follow.

LnL7 committed:
use relative path for module tests

LnL7 commented on lib/tests/modules/default.nix:
```diff
@@ -1,4 +1,4 @@
-{ lib ? import <nixpkgs/lib>, modules ? [] }:
+{ lib ? import ../.., modules ? [] }:
```
Is there a reason why this used `<nixpkgs/lib>`?

nbp replied:
I think this is a legacy of when I created the test suite, to check against old tree of Nixpkgs, with `-I nixpkgs=...`.

Nothing to worry about today.

LnL7 commented:
I added some basic tests and a better example for the manual. Anything else that's missing or should be improved?
@nbp Does the implementation using `modulesPath` look good to you or would you prefer something more like a searchpath for modules.

The question about disabling a single file vs the entire module closure is also still open. I don't really have an opinion about that.

nbp commented on lib/tests/modules.sh:
```diff
@@ -99,6 +99,12 @@ checkConfigOutput 'true' "$@" ./define-enable.nix ./define-loaOfSub-if-foo-enabl
 checkConfigOutput 'true' "$@" ./define-enable.nix ./define-loaOfSub-foo-if-enable.nix
 checkConfigOutput 'true' "$@" ./define-enable.nix ./define-loaOfSub-foo-enable-if.nix
 
+set -- config.enable ./define-enable.nix ./declare-enable.nix
```
nit: add a one line comment to describe what these tests are for.

nbp commented on lib/tests/disable-module.nix:
```diff
@@ -0,0 +1 @@
+{ ... }:
```
nit: `git commit -a` ? Remove this file.

nbp commented on lib/modules.nix:
```diff
@@ -58,10 +58,7 @@ rec {
-      # compatibility with the old module system.  Not sure if this is
-      # the most sensible policy.
-      options = mergeModules prefix (reverseList closed);
+      options = mergeModules prefix (filterModules (specialArgs.modulesPath or "") closed);
```
nit: This is adding a meaning to a name which used to be ignored by the module system.  Add a paragraph to the existing `specialArgs` comment to describe what is the intended meaning of `modulesPath`.

LnL7 replied:
added a comment to the definition of `evalModules`

nbp commented on lib/tests/modules/disable-define-enable.nix:
```diff
@@ -0,0 +1,5 @@
+{ lib, ... }:
+
+{
+  disabledModules = [ ./define-enable.nix ];
```
nit: Change `lib/tests/modules/default.nix` to add a `modulesPath` special argument, and use it to double check that we can lookup for modules as the string `"define-enable.nix"` instead of a path.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,48 @@
+
+<title>Replace Modules</title>
+
+<para>It's also possible to disable modules, removing the option
```
nit: Avoid starting new paragraph with `It`, maybe use the passive form.  I suggest reading the current documentation and try to mimic the current style to write these paragraph the same way.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,48 @@
+<para>It's also possible to disable modules, removing the option
+  declarations and config implementation defined within that file.
+  This can be used to import a set of modules from another channel
+  while keeping the rest of the system on a stable release.</para>
```
nit: This is actually a really valid use case.  I suggest you add an additional 4 lines example which just does that assuming that `nixos-rebuild` is used with an extra `-I custom-nixos=/home/dev/nixos-custom/nixos` command line argument.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,48 @@
+
+  cfg = config.services.postgresql;
+
+in
```
nit: remove these extra white-lines within the let-in.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,48 @@
+in
+
+{
+  disabledModules = [ "services/databases/postgresql.nix" ];
```
nit: Disabling a module implies that you at least declare the equivalent set of option, otherwise you will probably have error messages because options are defined but never declared.  I suggest you take a smaller existing module than `postgresql`.

nbp reviewed (approved):
This changes would look good with the following nits applied.

Minor documentation and test cases improvements to be fixed before merging. The documentation might require a second proof read once it is improved.

nit: add a change-log entry as well.

nbp commented:
> Apart from that, there is at least a design question which needs to be answer, which is: should we consider the closure of the excluded modules or only a single module at a time?

To answer my-self.  We only want to exclude a single module instead of the closure of modules, because the list of imports of a module can refer to common modules.

Counting the number of time a module is imported would be a nightmare to understand, and be a terrible interface.

Having to list all modules imported by the disabled modules would be painful but would at least be easy to understand as opposed to any automatic system decision.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,63 @@
+  import a set of modules from another channel while keeping the rest
+  of the system on a stable release.</para>
+<para><literal>disabledModules</literal> is a top level attribute like
+  <literal>options</literal> and <literal>config</literal>.  It
```
follow-up: mention `imports`, as `options` and `config` are both valid arguments of modules.

follow-up: I would not be surprised that there is another part of the documentation which list all these special attribtues, in which case we should add `disableModules` into it.

LnL7 replied:
There's not really an overview except for the [writing modules](http://nixos.org/nixos/manual/#sec-writing-modules) section. I could add it there, but I'm not sure if that's the best place.

nbp commented on nixos/doc/manual/development/replace-modules.xml:
```diff
@@ -0,0 +1,63 @@
+{ config, lib, pkgs, ... }:
+
+{
+  disabledModules = [ "services/databases/postgresql.nix" ];
```
nit: Add an introduction paragraph explaining each example.

LnL7 replied:
done

LnL7 commented:
@nbp ❤️ thank you for all the help with this.

nbp commented:
@globin , this change is marked as 17.03, we should either cherry-pick this change or move the release note to the next version.

globin commented:
I think I'd prefer not including this in 17.03 if you don't have strong feelings, to give this a period of time on master first.

LnL7 commented:
I personally would really like see this in 17.03 because it enables more flexibility when using a stable release without having to maintain a fork of nixpkgs.
That being said it's often possible to work around it in some other way, like mkForce or renaming the module from unstable.

nbp commented:
Reverted by commit fcec3e1c72ad7ac7eaacd00c8fc87dc04696fdb0

nbp commented:
re-made by commit d88721e.

basvandijk commented:
I would also prefer to see this in 17.03. In the documentation we can specify this is an experimental feature and that it's API might change in the future or that it might be removed. It's probably only going to be used by advanced users who know how to deal with changing API's.

Infinisil commented:
Note that with https://github.com/NixOS/nixpkgs/pull/76857 `disabledModules` now works recursively, meaning any `imports` by any disabled modules are ignored.
